### PR TITLE
MAJOR windows fixes for secretgen and antrea-cleanup

### LINKED ISSUE #988: Windows nodes lose connectivity on reboot
**Bug description**

We need the solution from https://github.com/kubernetes-sigs/sig-windows-tools/issues/177 to be pulled and run inside our YTT Templates.  This can easily be automated into image builder, but our windows kubelets should for the time being be able to wget and install this file on the fly if not available.

**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [ ] Addons
- [ ] CLI
- [ ] Docs
- [ ] IAM
- [x] Installation
- [ ] Plugin
- [ ] Security
- [ ] Test and Release
- [ ] User Experience

**Expected behavior**

Windows nodes, on reboot, dont lose connectivity

**Steps to reproduce the bug**

Run our windows recipes, make a working windows kubelet, and reboot it.

*note* this is a temporary  kubeadm bootstrapping workaround until antrea 1.4+ comes out which solve this.

### PR DESCRIPTION
**Pull request by hxietkg**
several fixes to windows

- regression for prod clusters fixed by skipping multi AZs
- fix to antrea  start after Windows nodes reboot because OVS bridge and HnsNetwork previously created by Antrea is not cleaned, a cleanup script is provided in the new version of Antrea, this change registers the
script as a callback on system's shutdown, the cleanup script will be invoked before Windows shutdown.
- fix to secret-gen does not support Windows cluster, just disable it for Windows

…

multiple-azs is supported by https://github.com/vmware-tanzu/tanzu-framework/pull/1108, but the change breaks Windows cluster, and this feature was not estimated for Windows cluster, so just disable it for Windows .

### What this PR does / why we need it
Antrea-agent fails to start after reboot, so the a Windows node will be re-created once it reboots, this is not expected, we need Windows node support reboot.

also, we fix the other issues mentioned above here for expedience 
### Which issue(s) this PR fixes

Fixes #988

### Describe testing done for PR
Create a windows cluster, then reboot Windows node, and verify antrea-agent in the node after reboot.

### Release note

```release-note
Windows nodes will support reboot.
```

### PR Checklist

- [ ] Squash the commits into one or a small number of logical commits
      
- [ ] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [ ] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

**Comment by jayunit100** [comment]
before merging lets add the one liner to `IS_WINDOWS_WORKLOAD_CLUSTER` as well to simplify the merge process /review process.  

Lets also rename this to "Minor windows fixes for secretgen and antreacleanup"

**Comment by jayunit100** [comment]
@hxietkg can you rename this to "two minor fixes to windows YTT" and add unit tests to the MR also (if its possible, I think it should be, I think we should be generating the secretgen in the yaml).

**Comment by hxietkg** [comment]
> @hxietkg can you rename this to "two minor fixes to windows YTT" and add unit tests to the MR also (if its possible, I think it should be, I think we should be generating the secretgen in the yaml).

this is not a bug, we may just disable the secret-gen by adding "SECRETGEN_CONTROLLER_ENABLE: false" to config file.

**Comment by jayunit100** [comment]
thats fair can you add that as a DEFAULT to the windows ytt then ?

**Comment by jayunit100** [comment]
hiya hongsheng , can you update the YTT tests like so:

```
	BeforeEach(func() {
		paths = []string{
				filepath.Join(yamlRoot, "config_default.yaml"),
				filepath.Join("./fixtures/tkr-bom-v1.21.1.yaml"),
				filepath.Join("./fixtures/tkg-bom-v1.4.0.yaml"),
				filepath.Join(yamlRoot, "infrastructure-vsphere", "v1.0.1", "ytt", "base-template.yaml"),
				filepath.Join(yamlRoot, "infrastructure-vsphere", "v1.0.1", "ytt", "overlay-windows.yaml"),
				filepath.Join(yamlRoot, "ytt", "02_addons", "cni", "antrea", "antrea_addon_data.lib.yaml"),
				filepath.Join(yamlRoot, "ytt", "02_addons", "cpi", "cpi_addon_data.lib.yaml"),
				filepath.Join(yamlRoot, "ytt", "03_customizations", "02_avi", "ako-deployment.lib.yaml"),
				filepath.Join(yamlRoot, "ytt/03_customizations/03_windows/prevent_windows_updates.yaml"),
				filepath.Join(yamlRoot, "ytt/03_customizations/annotate_os_info.yaml"),
				filepath.Join(yamlRoot, "ytt","03_customizations","registry_skip_tls_verify.yaml"),
				filepath.Join(yamlRoot,"ytt","03_customizations","registry_ca_cert.yaml"),
				//filepath.Join(YAML_ROOT, "provider-bundle", "providers", "ytt", "02_addons", "cpi", "cpi_addon_data.lib.yaml"),
				filepath.Join(yamlRoot, "ytt"), // lib/helpers.star, lib/config_variable_association.star, lib/validate.star

		}
```

File: pkg/v1/providers/tests/unit/windows_test.go [hunk 1]
```diff
@@ -30,6 +30,7 @@ var _ = Describe("Windows Ytt Templating", func() {
 			filepath.Join(yamlRoot, "ytt", "02_addons", "cpi", "cpi_addon_data.lib.yaml"),
 			filepath.Join(yamlRoot, "ytt", "03_customizations", "02_avi", "ako-deployment.lib.yaml"),
 			//filepath.Join(YAML_ROOT, "provider-bundle", "providers", "ytt", "02_addons", "cpi", "cpi_addon_data.lib.yaml"),
+			filepath.Join(yamlRoot, "ytt", "03_customizations", "03_windows"),
```
Comment (jayunit100) [inline pkg/v1/providers/tests/unit/windows_test.go:33]: ah yes ! thank you for adding this.

File: pkg/v1/providers/ytt/03_customizations/03_windows/prevent_windows_updates.yaml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 #@ load("@ytt:overlay", "overlay")
 #@ load("@ytt:data", "data")
 #@ if data.values.IS_WINDOWS_WORKLOAD_CLUSTER:
-#@overlay/match by=overlay.subset({"kind":"KubeadmConfigTemplate"})
+#@overlay/match by=overlay.subset({"kind":"KubeadmConfigTemplate", "metadata":{"name": str(data.values.CLUSTER_NAME) + "-md-0-windows-containerd"}})
```
Comment (jayunit100) [inline pkg/v1/providers/ytt/03_customizations/03_windows/prevent_windows_updates.yaml:4]: interesting . I assume this works, but am curious what changed that now requires us to start doing this?

**Comment by hxietkg** [comment]
rebased on 15/Nov

File: pkg/v1/providers/infrastructure-vsphere/v1.0.1/ytt/base-template.yaml [hunk 3]
```diff
@@ -65,7 +65,55 @@ spec:
 apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
 kind: VSphereMachineTemplate
 metadata:
-  name: '${ CLUSTER_NAME }-worker'
```
Comment (jayunit100) [inline pkg/v1/providers/infrastructure-vsphere/v1.0.1/ytt/base-template.yaml:68]: well need to potentially disable `VSPHERE_AZ_0` stuff for now, just to get this merged and working for release

Comment (jayunit100) [reply pkg/v1/providers/infrastructure-vsphere/v1.0.1/ytt/base-template.yaml:68]: possible solutions:
- disable VSPHERE_AZs
- enable AZ1/2 in our windows ytt

going w/ one bc...well... *time* :)

Comment (jayunit100) [reply pkg/v1/providers/infrastructure-vsphere/v1.0.1/ytt/base-template.yaml:68]: @stuartpreston fYi were disabling this for expedeince

**Comment by vuil** [comment]
Please add an entry in the release notes section.
Also please update the PR descriptoin since there is more than just introducing the cleanup script in the change now.

**Comment by jayunit100** [comment]
Lgtm 

clustergen tests are correctly reporting removal of the old error 

“””Error: : unable to get template: Overlaying (in following order: overlay-windows.yaml, vsphere-overlay.yaml, 01_plans/prod.yaml,”””

**Comment by jayunit100** [comment]
Hongsheng please add release note

**Review by vuil — APPROVED** [review]
lgtm, thanks!

**Pull request merged**